lib/pharos/phases/configure_master.rb

```diff
@@ -118,36 +134,108 @@ def generate_config
 
         # Only configure etcd if the external endpoints are given
         if @config.etcd&.endpoints
-          config['etcd'] = {
-            'endpoints' => @config.etcd.endpoints
-          }
-
-          config['etcd']['certFile'] = '/etc/pharos/etcd/certificate.pem' if @config.etcd.certificate
-          config['etcd']['caFile'] = '/etc/pharos/etcd/ca-certificate.pem' if @config.etcd.ca_certificate
-          config['etcd']['keyFile'] = '/etc/pharos/etcd/certificate-key.pem' if @config.etcd.key
+          configure_external_etcd(config)
+        else
+          configure_internal_etcd(config)
         end
 
         config['apiServerExtraArgs'] = {}
         config['apiServerExtraVolumes'] = []
 
         # Only if authentication token webhook option are given
-        if @config.authentication&.token_webhook
-          config['apiServerExtraArgs'].merge!(authentication_token_webhook_args(@config.authentication.token_webhook.cache_ttl))
-          config['apiServerExtraVolumes'] += volume_mounts_for_authentication_token_webhook
-        end
+        configure_token_webhook(config) if @config.authentication&.token_webhook
 
         # Configure audit related things if needed
-        if @config.audit&.server
-          config['apiServerExtraArgs'].merge!(
-            "audit-webhook-config-file" => AUDIT_CFG_DIR + '/webhook.yml',
-            "audit-policy-file" => AUDIT_CFG_DIR + '/policy.yml'
-          )
-          config['apiServerExtraVolumes'] += volume_mounts_for_audit_webhook
-        end
+        configure_audit_webhook(config) if @config.audit&.server
 
         config
       end
 
+      # @return [Array<String>]
+      def build_extra_sans
+        extra_sans = Set.new(['127.0.0.1'])
+        @config.master_hosts.each do |h|
+          extra_sans += [h.address, h.private_address].compact.uniq
+        end
+
+        extra_sans.to_a
+      end
+
+      # Copies certificates from leading master
+      def copy_kube_certs
+        leader = @config.master_leader
+        leader_ssh = Pharos::SSH::Client.for_host(leader)
```

Comment:
Either needs a separate phase for running `kubeadm alpha phase certs` on the leader, or then run the `ConfigureMaster` phase first on the leader, then grab the generated certs somewhere locally before running the phase on the follower nodes to deploy the existing CA/SA certs.